include/rocksdb/experimental.h

```diff
@@ -61,106 +61,285 @@ Status UpdateManifestForFilesState(
+// concatenated sequence of segments from i to j (inclusive) compares between
+// b1 and b2 according to comparator c.
+//    EXAMPLE: To represent that segment of ordinal i is equal to s, that would
+//    be <i,i,bytewise_comparator,before(s),after(s)>.
+//    NOTE: To represent something like segment has a particular prefix, you
+//    would need to split the key into more segments appropriately. There is
+//    little loss of generality because we can combine adjacent segments for
+//    specifying where clauses and implementing filters.
+// The DIRECT FILTER for a where clause is an order-based filter on the same
+// sequence of segments and comparator (non-empty intersection of bounds/keys),
+// or in the special case of an equality clause (see example), an equivalence
+// class filter on the sequence of segments.
+//
+// GENERALIZING FILTERS (INDIRECT):
+// * Point queries can utilize essentially any kind of filter by extracting
+// applicable segments of the query key (if not using whole key) and querying
+// the corresponding equivalence class or trivial range.
+//    NOTE: There is NO requirement e.g. that the comparator used by the filter
+//    match the CF key comparator or similar. The extractor simply needs to be
+//    a pure function that does not return "out of bounds" segments.
+//
+// Beyond point queries, we generally expect the key comparator to be a
+// lexicographic / big endian ordering at a high level, while each segment
+// can use an arbitrary comparator.
+//
+// * Equivalence class filters can apply to range queries under conditions
+// resembling legacy prefix filtering (prefix_extractor). An equivalence class
+// filter on segments i through j and category set s is applicable to a range
+// query from lb to ub if
+//   * All segments through j extracted from lb and ub are equal.
+//     NOTE: being in the same filtering equivalence class is insufficient, as
+//     that could be unrelated inputs with a hash collision. Here we are
+//     omitting details that would formally accommodate comparators in which
+//     different bytes can be considered equal.
+//   * The categories of lb and ub are in the category set s.
+//   * COMMON SEGMENT PREFIX PROPERTY (for all x, y, z; params j, s): if
+//     * Keys x and z have equal segments up through ordinal j, and
+//     * Keys x and z are in categories in category set s, and
+//     * Key y is ordered x < y < z according to the CF comparator,
+//   then both
+//     * Key y has equal segments up through ordinal j (compared to x and z)
+//     * Key y is in a category in category set s
+//  (This is implied by the SEGMENT MAXIMAL PREFIX PROPERTY in the simplified
+//  model.)
+//
+// * Order-based filters on segments (rather than whole key) can apply to range
+// queries (with "whole key" bounds). Specifically, an order-based filter on
+// segments i through j and category set s is applicable to a range query from
+// lb to ub if
```

Comment:
nit:`lb` and `ub` also both need to be in category set s, right?

Reply:
I forgot to mention that this time! Yes, thanks!